### PR TITLE
3.0: Amend/Add Schema: compute resource, custom actions, iam, image, dns

### PR DESCRIPTION
**Pull request by hanwen-pcluste**
1. Add `Name` field to compute resource
2. Amend and move custom actions to headnode and queue of slurm and headnode of awsbatch
3. Add IAM support in headnode and queue
4. Add image support in headnode and queue
5. Add DNS settings support in slurm settings
6. classes in cluster.py and cluster_schema.py are reordered to have node and cluster related classes defined at last so that we can use IAM for nodes.

ToDo
Although IAM section has a minor difference depending on if it is on global level or lower level, I am using the same classes to make the code concise. The same idea applies to image section. We need to revisit to see whether we can change the design of the configuration file or we have to write more code for it.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/tests/pcluster/schemas/test_schema_validators.py [hunk 1]
```diff
@@ -154,8 +154,16 @@ def test_awsbatch_compute_resource_validator(section_dict, expected_message):
     "section_dict, expected_message",
     [
         ({"ComputeResources": []}, "Length must be 1"),
-        ({"ComputeResources": [{"InstanceType": "c5.xlarge"}]}, None),
-        ({"ComputeResources": [{"InstanceType": "c5.xlarge"}, {"InstanceType": "c4.xlarge"}]}, "Length must be 1"),
+        ({"ComputeResources": [{"Name": "compute_resource1", "InstanceType": "c5.xlarge"}]}, None),
+        (
+            {
+                "ComputeResources": [
+                    {"Name": "compute_resource1", "InstanceType": "c5.xlarge"},
```
Comment (enrico-usai) [inline cli/tests/pcluster/schemas/test_schema_validators.py:161]: I think we should re-use the QueueNameValidator for the compute name too or put some code in common.

I don't think we can have any value for the compute resource name.

File: cli/src/pcluster/models/cluster.py [hunk 2]
```diff
@@ -663,7 +553,122 @@ def _register_validators(self):
...
+        self.custom_ami = Resource.init_param(custom_ami)
+
+
+class CustomAction(Resource):
+    """Represent a custom action resource."""
+
+    def __init__(self, script: str, args: List[str] = None, event: str = None, run_as: str = None):
+        super().__init__()
+        self.script = Resource.init_param(script)
+        self.args = args
```
[the remark this answers is not in the document]
Comment (hanwen-pcluste) [reply cli/src/pcluster/models/cluster.py]: Done

File: cli/src/pcluster/models/cluster_slurm.py [hunk 3]
```diff
@@ -108,12 +104,22 @@ def instance_type_list(self):
         return [compute_resource.instance_type for compute_resource in self.compute_resources]
 
 
+class Dns(Resource):
+    """Represent the DNS settings."""
+
+    def __init__(self, disable_managed_dns: bool = None, domain: str = None, hosted_zone_id: str = None):
+        super().__init__()
+        self.disable_managed_dns = Resource.init_param(disable_managed_dns, default=False)
+        self.domain = domain
```
[the remark this answers is not in the document]
Comment (hanwen-pcluste) [reply cli/src/pcluster/models/cluster_slurm.py]: Done

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 4]
```diff
@@ -365,16 +521,33 @@ def make_resource(self, data, **kwargs):
...
     instance_type = fields.Str(required=True)
     simultaneous_multithreading = fields.Bool()
     networking = fields.Nested(HeadNodeNetworkingSchema, required=True)
+    image = fields.Nested(ImageSchema)
     ssh = fields.Nested(SshSchema, required=True)
     storage = fields.Nested(StorageSchema)
     dcv = fields.Nested(DcvSchema)
     efa = fields.Nested(EfaSchema)
+    custom_actions = fields.Nested(CustomActionSchema, many=True)
+    iam = fields.Nested(IamSchema, data_key="IAM")
```
[the remark this answers is not in the document]
Comment (hanwen-pcluste) [reply cli/src/pcluster/schemas/cluster_schema.py]: Done

File: cli/tests/pcluster/schemas/test_schema_validators.py (hunk 1 above)
Comment (hanwen-pcluste) [reply cli/tests/pcluster/schemas/test_schema_validators.py:161]: Done, any reason to refuse "default" as a name?

**Review by enrico-usai — APPROVED** [review]
LGTM

**Pull request merged**